clients/js/genjs.go

```diff
@@ -0,0 +1,249 @@
+		Version:     s.Info.InfoProps.Version,
+		Description: s.Info.InfoProps.Description,
+	}
+
+	for _, path := range swagger.SortedPathItemKeys(s.Paths.Paths) {
+		pathItem := s.Paths.Paths[path]
+		pathItemOps := swagger.PathItemOperations(pathItem)
+		for _, method := range swagger.SortedOperationsKeys(pathItemOps) {
+			op := pathItemOps[method]
+			if op.Deprecated {
+				continue
+			}
+			methodCode, err := methodCode(op, method, s.BasePath, path)
+			if err != nil {
+				return err
+			}
+			tmplInfo.Methods = append(tmplInfo.Methods, methodCode)
+		}
+	}
+
+	indexJS, err := templates.WriteTemplate(indexJSTmplStr, tmplInfo)
+	if err != nil {
+		return err
+	}
+
+	packageJSON, err := templates.WriteTemplate(packageJSONTmplStr, tmplInfo)
+	if err != nil {
+		return err
+	}
+
+	if err = ioutil.WriteFile(filepath.Join(modulePath, "index.js"), []byte(indexJS), 0644); err != nil {
+		return err
+	}
+
+	if err = ioutil.WriteFile(filepath.Join(modulePath, "package.json"), []byte(packageJSON), 0644); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type clientCodeTemplate struct {
+	PackageName string /* Name of the npm package */
+	ClassName   string /* Name of the class to export */
+	ServiceName string /* Service's internal name (aka the github repo/discovery name) */
+	Version     string
+	Description string
+	Methods     []string
+}
+
+var indexJSTmplStr = `
+const discovery = require("@clever/discovery");
+const request = require("request");
+const url = require("url");
+const opentracing = require("opentracing");
+
+// go-swagger treats handles/expects arrays in the query string to be a string of comma joined values
+// so...do that thing. It's worth noting that this has lots of issues ("what if my values have commas in them?")
+// but that's an issue with go-swagger
+function serializeQueryString(data) {
+  if (Array.isArray(data)) {
+    return data.join(",");
+  }
+  return data;
+}
+
+module.exports = class {{.ClassName}} {
+
+  address;
+  timeout;
+
+  constructor(options) {
+    options = options || {};
+
+    if (options.discovery) {
+      this.address = discovery("{{.ServiceName}}", "default").url();
+    } else if (options.address) {
+      this.address = options.address;
+    } else {
+      throw new Error("Cannot initialize {{.ServiceName}} without discovery or address");
+    }
+    if (options.timeout) {
+      this.timeout = options.timeout
+    }
+  }
+{{range $methodCode := .Methods}}{{$methodCode}}{{end}}
+}
+`
+
+var packageJSONTmplStr = `
+{
+  "name": "{{.PackageName}}",
+  "version": "{{.Version}}",
+  "description": "{{.Description}}",
+  "main": "index.js",
+  "dependencies": {
+    "@clever/discovery": "0.0.8",
+    "opentracing": "^0.11.1",
+    "request": "^2.75.0"
+  }
+}
+`
+
+var methodTmplStr = `
+  {{.MethodName}} = ({{range $param := .Params}}{{$param.JSName}}, {{end}}options, cb) => {
+    if (!cb && typeof options === "function") {
+      cb = options;
+      options = undefined;
+    }
```

Comment:
I think the discussion of if Clever supports callbacks makes a lot of sense to have, but I strongly feel it is not in the scope of this repo to impart any push one way or another. _Today_ Clever uses both, and we can trivially support both in the library, _or_ we can push the burden onto all our consumers to use bluebird etc until the day when we add a lint rule that says "no callbacks ever". 

Before that day we're just making the experience of using wag clients worse for no real benefit. Take this code in the `il-api` - this is unnecessarily complex, since we are mixing callbacks and promises. https://github.com/Clever/instant-login-api/blob/master/apps/permissions.coffee#L38-L52

I'm all for a plan to fully move off of callbacks, but I'm not a fan of using these generated clients as a blunt to help us get there.